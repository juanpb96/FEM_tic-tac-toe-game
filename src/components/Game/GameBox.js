import { useState } from 'react';

export const GameBox = ({ player }) => {
    const [isEmpty, setIsEmpty] = useState(true);
    const [mark, setMark] = useState({
        icon: '',
        player: ''
    });
    
    const handlePlayerClick = () => {
        if (!isEmpty) {
            return;
        }

        setIsEmpty(false);

        const icon = player === 'X'
            ? `${process.env.PUBLIC_URL}/assets/icon-x.svg`
            : `${process.env.PUBLIC_URL}/assets/icon-o.svg`;

        setMark({ 
            icon,
            player
        });
    };

    return (
        <button onClick={ handlePlayerClick }>
            { !isEmpty && <img src={ mark.icon } alt={ mark.player } /> }
        </button>
    );
};
