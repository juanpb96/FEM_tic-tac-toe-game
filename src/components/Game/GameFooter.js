import { STORAGE } from '../../types/types';

const {
    lsCpuMark,
    lsCpuScore,
    lsPlayerMark,
    lsPlayerScore,
    lsP1Mark,
    lsP1Score,
    lsP2Mark,
    lsP2Score,
    lsTiedScore,
} = STORAGE;

export const GameFooter = () => {
    const cpuMark = localStorage.getItem(lsCpuMark);
    const cpuScore = localStorage.getItem(lsCpuScore);
    const p1Mark = localStorage.getItem(cpuMark ? lsPlayerMark : lsP1Mark);
    const p1Score = localStorage.getItem(cpuMark ? lsPlayerScore : lsP1Score);
    const p2Mark = localStorage.getItem(lsP2Mark);
    const p2Score = localStorage.getItem(lsP2Score);
    const tiedScore = localStorage.getItem(lsTiedScore);

    let labelPlayerX = '';
    let labelPlayerO = '';
    let labelScoreX = 0;
    let labelScoreO = 0;
    
    switch ('X') {
        case cpuMark:
            labelPlayerX = 'CPU';
            labelPlayerO = 'YOU';
            labelScoreX = cpuScore;
            labelScoreO = p1Score;
            break;
        case p1Mark:
            labelPlayerX = cpuMark ? 'YOU' : 'P1';
            labelPlayerO = cpuMark ? 'CPU' : 'P2';
            labelScoreX = p1Score;
            labelScoreO = cpuScore;
            break;
        case p2Mark:
            labelPlayerX = 'P2';
            labelPlayerO = 'P1';
            labelScoreX = p2Score;
            labelScoreO = p1Score;
            break;
        default:
            break;
    }

    return (
        <footer>
            <section aria-label='score'>
                <h2>X { `(${labelPlayerX})` }</h2>
                <p>{ labelScoreX }</p>
            </section>
            <section aria-label='score'>
                <h2>TIES</h2>
                <p>{ tiedScore }</p>
            </section>
            <section aria-label='score'>
                <h2>O {`(${labelPlayerO})`}</h2>
                <p>{ labelScoreO }</p>
            </section>
        </footer>
    );
};
