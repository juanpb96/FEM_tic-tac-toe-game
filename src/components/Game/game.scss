@import '../../styles/mixins/media';

/* Header */
.turn {
    padding: 0.5625rem 0.9375rem 0.8125rem;
    line-height: 1.125rem;

    & img {
        width: 1rem;
    }
}

.button-restart {
    cursor: pointer;
    padding: 0.75rem;
    color: var(--semi-dark-navy);
    background-color: var(--silver);
    border: none;
    box-shadow: var(--shadow-silver-secondary);
    transition: color .5s ease-in-out,
                background-color .5s ease-in-out;

    &:where(:hover, :focus) {
        color: hsla(199, 35%, 19%, 1);
        background-color: var(--silver-active);
    }

    & svg {
        width: 1rem;
        height: 1rem;
    }
}

@include media(tablet) {
    .turn {
        padding: 0.8125rem 1.875rem 1.1875rem;
        line-height: 1.25rem;

        & img {
            width: 1.25rem;
        }
    }

    .button-restart {
        padding: 1rem;

        & svg {
            width: 1.25rem;
            height: 1.25rem;
        }
    }
}
/* End Header */

/* Modal */
.modal-container {
    position: absolute;
    inset: 0;
    z-index: 10;
    background-color: hsla(0, 0%, 0%, .5);
}

.modal {
    width: 100%;
    height: 14.25rem;

    .result {
        letter-spacing: 0.875px;
    }

    .title {
        font-size: clamp(1.5rem, 6vw, 2.5rem);

        & img {
            width: 1.75rem;
        }
    }
}

@include media(tablet) {
    .modal {
        height: 16.625rem;

        .title {
            & img {
                width: 4rem;
            }
        }
    }
}
/* End Modal */

/* Board */
// TODO: Add transitions to improve win state display

.board {
    display: grid;
    grid-template: repeat(3, 1fr) / repeat(3, 1fr);
    gap: 1.25rem;

    &-box {
        width: min(100%, 8.75rem);
        aspect-ratio: 1 / 1;
        padding: 0;
        position: relative;
        border: none;
        
        &:empty {
            cursor: pointer;
            position: relative;

            &:where(:hover, :focus) {
                &::before {
                    content: '';
                    position: absolute;
                    inset: 0;
                    aspect-ratio: 1 / 1;
                    // TODO: Use a mixin to reuse code below
                    width: clamp(2.5rem, 10vw, 4rem);
                    margin: 1.5rem auto 2rem;
                    // --------------------------
                    background-size: cover;
                }

                &[data-turn='X'] {
                    &::before {
                        background-image: url(../../styles/assets/icon-x-outline.svg);
                    }
                }

                &[data-turn='O'] {
                    &::before {
                        background-image: url(../../styles/assets/icon-o-outline.svg);
                    }
                }
            }
        }

        &.X-box {
            color: var(--light-blue);

            &[data-win='true'] {
                color: var(--dark-navy);
                background-color: var(--light-blue);
                box-shadow: var(--shadow-blue-primary);
            }

        }
        
        &.O-box {
            color: var(--light-yellow);

            &[data-win='true'] {
                color: var(--dark-navy);
                background-color: var(--light-yellow);
                box-shadow: var(--shadow-yellow-primary);
            }
        }

        & svg {
            width: clamp(2.5rem, 10vw, 4rem);
            height: auto;
            margin: 1.5rem auto 2rem;
        }
    }
}

@include media(tablet) {
    .board {
        &-box {
            &:empty {
                &:where(:hover, :focus) {
                    &::before {
                        margin: 2.375rem auto;
                    }
                }
            }
        }

        & svg {
            margin: 2.375rem auto;
        }
    }
}
/* End Board */

/* Footer */
.score {
    & > * {
        line-height: 1.25;
    }
}

@include media(tablet) {
    .score {
        & > * {
            line-height: 1.26;
        }
    }
}
/* End Footer */