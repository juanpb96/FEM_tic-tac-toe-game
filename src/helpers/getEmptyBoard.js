const EMPTY_BOARD = [
    [null, null, null],
    [null, null, null],
    [null, null, null],
];

export const getEmptyBoard = () => EMPTY_BOARD.map(row => row.slice(0));