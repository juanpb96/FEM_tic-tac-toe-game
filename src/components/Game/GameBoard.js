import { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { PlayerCoxtext } from '../../hocs/PlayerContext';
import { MODAL_TYPES, STORAGE } from '../../types/types';
import { GameBox } from './GameBox';

const {
    lsBoardState,
    lsTurnCount,
    lsPlayerMark,
    lsCpuMark,
    lsP1Mark,
    lsP2Mark,
    lsCurrentTurnMark,
    lsCpuScore,
    lsPlayerScore,
    lsP1Score,
    lsP2Score,
    lsTiedScore,
} = STORAGE;

const USER = {
    player: 'player',
    p1: 'p1',
    p2: 'p2,',
    cpu: 'cpu',
};

export const GameBoard = ({ setModalValues, setShowModal }) => {
    const { player, setPlayer } = useContext(PlayerCoxtext);

    // TODO: When match ends and player leaves.
    //       Display an empty board when player returns

    const emptyBoard = [
        [null, null, null],
        [null, null, null],
        [null, null, null],
    ];

    let initialBoard = localStorage.getItem(lsBoardState);
    initialBoard = initialBoard && JSON.parse(initialBoard);

    const [board, setBoard] = useState(initialBoard || emptyBoard);

    const boardRef = useRef(null);
    const isGameOver = useRef(false);
    const turnCounter = useRef(+localStorage.getItem(lsTurnCount) || 0);

    const cpuMark = localStorage.getItem(lsCpuMark);
    const searchFirstMove = board.some(row => row.some(cell => cell === cpuMark));
    const isCpuFirstMove = useRef(!searchFirstMove);

    let p1Mark;

    if (cpuMark) {
        p1Mark = localStorage.getItem(lsPlayerMark);
    }

    const updateBoard = (row, col) => {
        setBoard(() => {
            board[row][col] = player;
            return board;
        });
        setPlayer(player === 'X' ? 'O' : 'X');
        turnCounter.current += 1;

        console.log(board);
    };

    const updateScore = (winner = '') => {
        switch (winner) {
            case USER.cpu:
                const cpuCurrentScore = +localStorage.getItem(lsCpuScore);
                localStorage.setItem(
                    lsCpuScore,
                    cpuCurrentScore ? cpuCurrentScore + 1 : 1
                );
                return;
            case USER.player:
                const playerCurrentScore = +localStorage.getItem(lsPlayerScore);
        
                localStorage.setItem(
                    lsPlayerScore,
                    playerCurrentScore ? playerCurrentScore + 1 : 1
                );
                return;
            default:
                const tieCurrentScore = +localStorage.getItem(lsTiedScore);
        
                localStorage.setItem(
                    lsTiedScore,
                    tieCurrentScore ? tieCurrentScore + 1 : 1
                );
                break;
        }
    };

    const clickBox = useCallback(
        (x, y) => {
            const buttons = boardRef.current.getElementsByTagName('button');
    
            console.log('Click', {x, y});
    
            switch (x) {
                case 0:
                    buttons[y].click();
                    break;
                case 1:
                    buttons[x + y + 2].click();
                    break;
                case 2:
                    buttons[x + y + 4].click();
                    break;
                default:
                    makeRandomMove();
                    break;
            }
        },
        [],
    );

    const displayModal = (messageType, mark) => {   
        setModalValues(prev => ({
            ...prev,
            type: messageType,
            winnerMark: mark,
        }));
        setShowModal(true);
    };

    const checkWinCondition = (p1Marks, p2Marks, isCellEmpty = false) => {
        // FIXME: Modify this when implementing logic for 2 players

        if (p1Marks < 3 && p2Marks < 3 && !(isCellEmpty && p2Marks === 2)) {
            return false;
        }

        console.log('GAME OVER', { p1Marks, p2Marks, isCellEmpty });

        isGameOver.current = true;

        if (p1Marks === 3) {
            displayModal(MODAL_TYPES.player_won, p1Mark);
            updateScore(USER.player);
        } 
        
        if (p2Marks === 3 || p2Marks === 2) {
            displayModal(MODAL_TYPES.player_lost, cpuMark);
            updateScore(USER.cpu);
        }

        return true;
    };

    const makeCpuDiagonalMove = ({ playerOcurrences, cpuOcurrences, isCellEmpty, x, y }) => {
        if (checkWinCondition(playerOcurrences, cpuOcurrences, isCellEmpty)) {
            if (isCellEmpty && cpuOcurrences === 2) {
                clickBox(x, y);
            }

            return true;
        }

        if (isCellEmpty && playerOcurrences === 2) {
            clickBox(x, y);
            return true;
        }

        return false;
    };

    const updateBoardValues = (cell, boardValues, x, y) => {
        switch (cell) {
            case p1Mark:
                boardValues.playerOcurrences++;
                break;
            case cpuMark:
                boardValues.cpuOcurrences++;
                break
            case null:
                boardValues.isCellEmpty = true;
                boardValues.x = x;
                boardValues.y = y;
                break;
            default:
                break;
        }
    };

    const checkBoardRowsCols = (type) => {
        const blockMove = {
            x: 0,
            y: 0,
            shouldBlock: false,
        };

        const boardValues = {
            x: 0,
            y: 0,
            playerOcurrences: 0,
            cpuOcurrences: 0,
            isCellEmpty: false,
        };

        for (let row = 0; row < board.length; row++) {
            boardValues.playerOcurrences = 0;
            boardValues.cpuOcurrences = 0;
            boardValues.isCellEmpty = false;

            for (let col = 0; col < board.length; col++) {
                if (type === 'rows') {
                    updateBoardValues(board[row][col], boardValues, row, col);
                    continue;
                } 
                    
                updateBoardValues(board[col][row], boardValues, col, row);
            }

            const { playerOcurrences, cpuOcurrences, isCellEmpty, x, y } = boardValues;

            if (checkWinCondition(playerOcurrences, cpuOcurrences, isCellEmpty)) {

                if (isCellEmpty && cpuOcurrences === 2) {
                    blockMove.x = x;
                    blockMove.y = y;
                    blockMove.shouldBlock = true;

                    return { ...blockMove };
                }

                return;
            }

            if (isCellEmpty && playerOcurrences === 2) {
                blockMove.x = x;
                blockMove.y = y;
                blockMove.shouldBlock = true;
            }
        }

        return { ...blockMove };
    };

    const makeCpuMove = useCallback(
        () => {
            // Check rows
            const blockRowCell = checkBoardRowsCols('rows');

            if (blockRowCell?.shouldBlock) {
                clickBox(blockRowCell.x, blockRowCell.y);
                return;
            }

            // Check columns
            const blockColCell = checkBoardRowsCols('cols');

            if (blockColCell?.shouldBlock) {
                clickBox(blockColCell.x, blockColCell.y);
                return;
            }

            const boardValues = {
                x: 0,
                y: 0,
                playerOcurrences: 0,
                cpuOcurrences: 0,
                isCellEmpty: false,
            };

            // Check diagonals
            // Top to bottom
            for (let i = 0; i < board.length; i++) {
                updateBoardValues(board[i][i], boardValues, i, i);
            }

            if (makeCpuDiagonalMove(boardValues)) {
                return;
            }
    
            // Bottom to top
            boardValues.playerOcurrences = 0;
            boardValues.cpuOcurrences = 0;
            boardValues.isCellEmpty = false;
    
            for (let row = 2, col = 0; row >= 0; row--, col++) {   
                updateBoardValues(board[row][col], boardValues, row, col);
            }

            if (makeCpuDiagonalMove(boardValues)) {
                return;
            }
    
            makeRandomMove();
        },
        [board, p1Mark, cpuMark, clickBox],
    );

    const makeRandomMove = () => {
        const isBoardFull = board.every(row => row.every(cell => cell === 'X' || cell === 'O'));

        if (isBoardFull) {
            displayModal(MODAL_TYPES.tied, '');
            updateScore();
            isGameOver.current = true;
            return;
        }

        const buttons = boardRef.current.getElementsByTagName('button');

        let pos = Math.floor(Math.random() * buttons.length);
        
        while (buttons[pos].hasChildNodes()) {
            pos = Math.floor(Math.random() * buttons.length);
        }

        console.log('Random Move', pos);
        buttons[pos].click();
    };

    useEffect(() => {
        if (isCpuFirstMove.current && cpuMark === player) {
            isCpuFirstMove.current = false;
            makeRandomMove();
            return;
        }

        if (cpuMark === player && !isGameOver.current) {
            makeCpuMove();
        }
    }, [cpuMark, player, makeCpuMove]);

    useEffect(() => {
        localStorage.setItem(lsBoardState, JSON.stringify(board));
        localStorage.setItem(lsCurrentTurnMark, player);
        localStorage.setItem(lsTurnCount, turnCounter.current);

        if (turnCounter.current === 10 && !isGameOver.current) {
            displayModal(MODAL_TYPES.tied, '');
            updateScore();
            isGameOver.current = true;
        }
    }, [board, player]);
    

    // TODO: Block user moves when game is over

    return (
        <main
            style={{
                display: 'grid',
                gridTemplate: 'repeat(3, 1fr) / repeat(3, 1fr)'
            }}
            ref={ boardRef }
        >
            {
                board.map(
                    (row, rowIndex) => row.map(
                        (cell, colIndex) => 
                            <GameBox
                                hasMark={ !!cell }
                                mark={ cell }
                                currentPlayer={ player }
                                row={ rowIndex }
                                col={ colIndex }
                                updateBoard={ updateBoard }
                                key={`${rowIndex}${colIndex}`} 
                            />
                    )
                )
            }
        </main>
    );
};
