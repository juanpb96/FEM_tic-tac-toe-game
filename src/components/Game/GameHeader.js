import { useContext } from 'react';

import { GameContext } from '../../hocs/GameContext';

import { ASSETS_PATH } from '../../helpers/constants';
import { MODAL_TYPES } from '../../types/types';

export const GameHeader = ({ openModal }) => {
    const { gameState: { currentPlayer } } = useContext(GameContext);

    const handleRestart = () => {
        openModal(MODAL_TYPES.restart);
    };

    return (
        <header className='flex space-between align-center mb-16'>
            <h1 className='sr-only'>Tic tac toe</h1>
            <img src={`${ASSETS_PATH}/logo.svg`} alt='' />
            <section
                className='[ turn ][ flex gap-2.5 box-navy br-1.5 ]'
                data-type='secondary'
            >
                <img src={`${ASSETS_PATH}/icon-turn-${currentPlayer}.svg`} alt={ currentPlayer } />
                <h2 className='fs-3.5 color-silver'>TURN</h2>
            </section>
            <button 
                aria-label='restart'
                className='[ button-restart ][ br-1.5 ]'
                data-type='secondary'
                onClick={ handleRestart }
            >
                <img src={`${ASSETS_PATH}/icon-restart.svg`} alt='' />
            </button>
        </header>
    );
};
