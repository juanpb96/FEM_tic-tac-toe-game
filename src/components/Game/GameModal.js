import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import { GameContext } from '../../hocs/GameContext';
import { ASSETS_PATH } from '../../helpers/constants';
import { ACTIONS, MODAL_TYPES, STORAGE } from '../../types/types';

const {
    lsBoardState,
    lsCurrentTurnMark,
    lsCpuMark,
    lsTurnCount,
} = STORAGE;

export const GameModal = ({ type, winnerMark, setShowModal }) => {
    const { dispatch } = useContext(GameContext);
    const navigate = useNavigate();

    const message = {
        title: 'TAKES THE ROUND',
        imgSrc: !winnerMark ? '' : `${ASSETS_PATH}/icon-${winnerMark}.svg`,
        result: '',
        button1: 'QUIT',
        button2: 'NEXT ROUND',
    };

    switch (type) {
        case MODAL_TYPES.restart:
            message.title = 'RESTART GAME?';
            message.button1 = 'NO, CANCEL';
            message.button2 = 'YES, RESTART';
            break;
        case MODAL_TYPES.tied:
            message.title = 'ROUND TIED';
            break;
        case MODAL_TYPES.player_won:
            message.result = 'YOU WON!';
            break;
        case MODAL_TYPES.player_lost:
            message.result = 'OH NO, YOU LOSTâ€¦';
            break;
        case MODAL_TYPES.player1_won:
            message.result = 'PLAYER 1 WINS!';
            break;
        case MODAL_TYPES.player2_won:
            message.result = 'PLAYER 2 WINS!';
            break;
        default:
            break;      
    }

    const { title, imgSrc, result, button1, button2 } = message;

    const handleClick = ({ target }) => {
        if (target.textContent === 'QUIT') {
            navigate('/new-game', { replace: true });
        }

        if (target.textContent === 'NEXT ROUND' || target.textContent === 'YES, RESTART') {
            localStorage.removeItem(lsBoardState);
            localStorage.setItem(lsCurrentTurnMark, 'X');
            localStorage.setItem(lsTurnCount, '1');
            
            dispatch({ type: ACTIONS.resetGame });

            if (localStorage.getItem(lsCpuMark)) {
                dispatch({
                    type: ACTIONS.setCpuMoveFirst,
                    payload: true
                });
            }
        }
        
        setShowModal(false);
    };

    return (
        <>
            { result && <p>{ result }</p> }
            <h2>
                { imgSrc && <img src={imgSrc} alt={winnerMark} /> }
                { title }
            </h2>
            <button onClick={ handleClick }>{ button1 }</button>
            <button onClick={ handleClick }>{ button2 }</button>
        </>
    );
};
