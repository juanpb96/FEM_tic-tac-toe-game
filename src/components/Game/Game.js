import { GameBoard } from './GameBoard';
import { GameFooter } from './GameFooter';
import { GameHeader } from './GameHeader';

export const Game = () => {
    return (
        <>
            <GameHeader />
            <GameBoard />
            <GameFooter />
        </>
    );
}
