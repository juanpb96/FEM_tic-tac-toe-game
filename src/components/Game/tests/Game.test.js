import { render, screen, fireEvent } from '@testing-library/react';

import { GameContext } from '../../../hocs/GameContext';
import { Game } from '../Game';

import { getEmptyBoard } from '../../../helpers/getEmptyBoard';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

const gameState = {
    currentPlayer: 'X',
    board: getEmptyBoard(),
    turnCounter: 1,
    isGameOver: false,
    isCpuFirstMove: true
};

const dispatch = jest.fn();

describe('Test <Game />', () => {
    test('should contain main title', () => { 
        render(
            <GameContext.Provider value={{
                gameState,
                dispatch
            }}>
                <Game />
            </GameContext.Provider>
        );

        expect(
            screen.getByRole('heading', { name: /Tic Tac Toe/i })
        ).toBeInTheDocument();
    });

    describe('modal should', () => { 
        test('open when user clicks on "restart" button and close if "No, cancel" is pressed', () => { 
            const { container } = render(
                <GameContext.Provider value={{
                    gameState,
                    dispatch
                }}>
                    <Game />
                </GameContext.Provider>
            );

            const restartButton = screen.getByRole('button', { name: 'restart' });

            fireEvent.click(restartButton);

            expect(screen.getByText('RESTART GAME?')).toBeInTheDocument();

            const cancelButton = screen.getByRole('button', { name: 'NO, CANCEL' });

            fireEvent.click(cancelButton);

            expect(container).toMatchSnapshot();
            expect(mockNavigate).not.toHaveBeenCalled();
        });
    });
});