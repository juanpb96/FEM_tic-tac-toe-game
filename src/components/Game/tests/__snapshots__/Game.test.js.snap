// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Game /> modal should open when user clicks on "restart" button and close if "No, cancel" is pressed 1`] = `
<div>
  <div
    class="[ page-content align-self-start mt-6 tablet:align-self-center tablet:mt-0 ]"
  >
    <header
      class="grid grid-cols-3 align-center gap-5 mb-16 tablet:mb-5"
    >
      <h1
        class="sr-only"
      >
        Tic tac toe
      </h1>
      <img
        alt=""
        src="/assets/logo.svg"
      />
      <section
        class="[ turn ][ flex gap-2.5 box-navy br-1.5 ]"
        data-type="secondary"
      >
        <img
          alt="X"
          src="/assets/icon-turn-X.svg"
        />
        <h2
          class="fs-3.5 color-silver tablet:fs-4"
        >
          TURN
        </h2>
      </section>
      <button
        aria-label="restart"
        class="[ button-restart ][ justify-self-end br-1.5 ]"
        data-type="secondary"
      >
        <svg
          fill="currentColor"
          height="20"
          role="presentation"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.524 0h-1.88a.476.476 0 0 0-.476.499l.159 3.284A9.81 9.81 0 0 0 9.835.317C4.415.317-.004 4.743 0 10.167.004 15.597 4.406 20 9.835 20a9.796 9.796 0 0 0 6.59-2.536.476.476 0 0 0 .019-.692l-1.348-1.349a.476.476 0 0 0-.65-.022 6.976 6.976 0 0 1-9.85-.63 6.987 6.987 0 0 1 .63-9.857 6.976 6.976 0 0 1 10.403 1.348l-4.027-.193a.476.476 0 0 0-.498.476v1.881c0 .263.213.476.476.476h7.944A.476.476 0 0 0 20 8.426V.476A.476.476 0 0 0 19.524 0Z"
            fill=""
          />
        </svg>
      </button>
    </header>
    <div
      class="[ modal-container ][ flex flex-center ]"
      data-animation="modal-closing"
    >
      <div
        class="[ modal ][ flex flex-center flex-col ]"
      >
        <h2
          class="[ title ][ flex flex-center gap-2.5 mb-6 fw-bold color-silver letter-m tablet:gap-6 tablet:letter-l tablet:mb-8 ]"
        >
          RESTART GAME?
        </h2>
        <div
          class="[ flex gap-4 ]"
        >
          <button
            class="[ button ][ silver ]"
            data-type="secondary"
          >
            NO, CANCEL
          </button>
          <button
            class="[ button ][ yellow ]"
            data-type="secondary"
          >
            YES, RESTART
          </button>
        </div>
      </div>
    </div>
    <main
      class="[ board ][ mb-5 ]"
    >
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
      <button
        aria-label="Empty"
        class="[ board-box  ][ box-navy br-2.5 ]"
        data-turn="X"
        data-type="primary"
        data-win="false"
      />
    </main>
    <footer
      class="[ flex space-between align-center gap-5 ]"
    >
      <section
        aria-label="score"
        class="[ score ][ flex flex-center flex-col flex-1 bg-light-blue p-3 br-2.5 ]"
      >
        <h2
          class="[ fs-3 fw-medium uppercase tablet:fs-3.5 ]"
        >
          X ()
        </h2>
        <p
          class="[ fs-5 fw-bold letter-s tablet:fs-6 tablet:letter-m ]"
        >
          0
        </p>
      </section>
      <section
        aria-label="score"
        class="[ score ][ flex flex-center flex-col flex-1 bg-silver p-3 br-2.5 ]"
      >
        <h2
          class="[ fs-3 fw-medium uppercase tablet:fs-3.5 ]"
        >
          TIES
        </h2>
        <p
          class="[ fs-5 fw-bold letter-s tablet:fs-6 tablet:letter-m ]"
        />
      </section>
      <section
        aria-label="score"
        class="[ score ][ flex flex-center flex-col flex-1 bg-light-yellow p-3 br-2.5 ]"
      >
        <h2
          class="[ fs-3 fw-medium uppercase tablet:fs-3.5 ]"
        >
          O ()
        </h2>
        <p
          class="[ fs-5 fw-bold letter-s tablet:fs-6 tablet:letter-m ]"
        >
          0
        </p>
      </section>
    </footer>
  </div>
</div>
`;
